### PR TITLE
Adafactor forloop basic impl

### PR DESCRIPTION
**Pull request by janeyx99**
#109581 

At this point, the vanilla implementation (the default) is good.
Docs: https://docs-preview.pytorch.org/pytorch/pytorch/129905/generated/torch.optim.Adafactor.html#torch.optim.Adafactor

Specifically, the impl in this PR, which attempts to replicate the paper,
```
optim = torch.optim.Adafactor([weight])
```
is close enough to https://pytorch-optimizers.readthedocs.io/en/latest/optimizer/#pytorch_optimizer.AdaFactor
```
optim_c = AdaFactor([weight], betas=(0, 0.999), scale_parameter=False)
```
is close enough to https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adafactor
```
optim = keras.optimizers.Adafactor(learning_rate=0.01)
```

The three results respectively for the same randomly generated weights:
```
# ours 
tensor([[ 0.3807594, -0.3912092],
        [ 0.0762539,  0.5377805],
        [ 0.2459473,  0.4662207]])

# pytorch-optimizer
tensor([[ 0.3807592, -0.3912172],
        [ 0.0762507,  0.5377818],
        [ 0.2459457,  0.4662213]])

# keras
array([[ 0.38076326, -0.39121315],
        [ 0.0762547 ,  0.5377859 ],
        [ 0.24594972,  0.46622536]], dtype=float32)
```

This gives me confidence to move forward in speeding up the implementation now that a baseline has been established. If you're curious about differences:
* keras assigns step_size (rho_t in their code) to `min(lr, 1 / sqrt(step)` whereas the OG impl uses a hardcoded 0.01 instead of lr. We do the same thing as keras, but our lr default is 0.01.
* We differ from the pytorch-optimizers default in that our default will not track momentum (thus `beta1=0`) and we do not apply parameter scaling.

<details>

Keras collab: https://colab.research.google.com/drive/1i3xF8ChL7TWKJGV_5v_5nMhXKnYmQQ06?usp=sharing

My script repro:

```
import torch
from pytorch_optimizer import AdaFactor
torch.set_printoptions(precision=7)

weight = torch.tensor([[ 0.37697506, -0.39500135],
        [ 0.07246649,  0.53399765],
        [ 0.24216151,  0.46243715]], dtype=torch.float32)
# bias = torch.tensor([0, 0], dtype=torch.float32)

weight.grad = torch.tensor([[-0.5940447, -0.7743838],
        [-0.5940447, -0.7743838],
        [-0.5940447, -0.7743838]], dtype=torch.float32)
# bias.grad = torch.tensor([-2.5027974,  1.5422692], dtype=torch.float32)

weight_c = weight.clone()
weight_c.grad = weight.grad.clone()

optim = torch.optim.Adafactor([weight])
optim.step()
print(weight)

optim_c = AdaFactor([weight_c], betas=(0, 0.999), scale_parameter=False)
optim_c.step()
print(weight_c)
```

<details>

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129905

### CODE REVIEW

File: torch/optim/adafactor.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+                exp_avg = torch.view_as_real(exp_avg)
+            param = torch.view_as_real(param)
+
+        # update step
+        step_t += 1
+
+        # Perform stepweight decay
+        param.mul_(1 - lr * weight_decay)
+
+        import math
```
Comment (vadimkantorov) [inline torch/optim/adafactor.py:237]: if `math` import is needed only for `sqrt(x)` - then maybe `x ** 0.5` can be used instead (same for `** 0.8`?) or is math.sqrt a bit faster?

Comment (janeyx99) [reply torch/optim/adafactor.py:237]: this is a draft! i don't plan on keeping this code lol

thanks for the suggestion

Comment (vadimkantorov) [reply torch/optim/adafactor.py:239]: there is also this :( https://stackoverflow.com/a/327011/445810 which says that `** 0.5` is a bit slower than math.sqrt. I hoped that `** 0.5` gets routed by python to math.sqrt under the hood...

Comment (janeyx99) [reply torch/optim/adafactor.py]: not in 3.10! 
image

File: torch/optim/adafactor.py [hunk 2]
```diff
@@ -248,9 +256,6 @@ def _single_tensor_adafactor(
             col_mean = torch.norm(grad, dim=-2).square_().div_(grad.size(-2))
             col_var.lerp_(col_mean, 1 - beta2_t)
             var_estimate = row_var.unsqueeze(-1) @ col_var.unsqueeze(-2)
-            # Include in docs later: we move epsilon1 to these 2 lines
-            # as an attempt to follow the intention of the paper authors
-            # in using eps1 for stability.
             var_est_denom = (
                 row_var.mean(dim=-1).clamp_(min=eps1).unsqueeze(-1).unsqueeze(-1)
```
Comment (vadimkantorov) [inline torch/optim/adafactor.py:260]: adding `keepdim=True` can save one of `unsqueeze` calls probably

File: torch/optim/adafactor.py (hunk 1 above)
Comment (vadimkantorov) [reply torch/optim/adafactor.py]: Also, for the same reasons dispatch_sqry handler can probably be eliminated in other optimizers

File: torch/optim/adafactor.py (hunk 2 above)
Comment (janeyx99) [reply torch/optim/adafactor.py]: Yea, keepdim is a good call for everywhere

File: docs/source/conf.py [hunk 3]
```diff
@@ -1751,6 +1751,8 @@
     "verify_aten_graph",
     # torch.optim.adadelta
     "adadelta",
+    # torch.optim.adafactor
+    "adafactor",
```
Comment (albanD) [inline docs/source/conf.py:1755]: No new entry should be added to this list. Please make sure they are properly documented on the website

File: torch/optim/adafactor.py [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+
+class Adafactor(Optimizer):
+    def __init__(
+        self,
+        params: ParamsT,
+        lr: Union[float, Tensor] = 1e-2,
+        beta2_decay: float = -0.8,
+        eps: Tuple[Optional[float], float] = (None, 1e-3),
+        d: float = 1,
+        weight_decay: float = 0,
```
Comment (albanD) [inline torch/optim/adafactor.py:25]:
```suggestion
        d: float = 1.,
        weight_decay: float = 0.,
```

File: torch/optim/adafactor.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+            d=d,
+            weight_decay=weight_decay,
+            maximize=maximize,
+        )
+        super().__init__(params, defaults)
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+        for group in self.param_groups:
+            group.setdefault("maximize", False)
```
Comment (albanD) [inline torch/optim/adafactor.py:54]: Do we need this? There cannot be an old checkpoint that is missing this flag right?

File: torch/optim/adafactor.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+    """
+)
+
+
+def _single_tensor_adafactor(
+    params: List[Tensor],
+    grads: List[Tensor],
+    row_vars: List[Tensor],
+    col_vars: List[Tensor],
+    variances: List[Tensor],
```
Comment (albanD) [inline torch/optim/adafactor.py:315]: I guess if you want to be nitpicky on types, these are List[Optional[Tensor]] not sure if it is valuable to present the difference though?

File: torch/optim/adafactor.py [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+    d: float,
+    lr: Union[Tensor, float],
+    beta2_decay: float,
+    weight_decay: float,
+    eps1: Optional[float],
+    eps2: float,
+    maximize: bool,
+    has_complex: bool,
+):
+    assert grad_scale is None and found_inf is None
```
Comment (albanD) [inline torch/optim/adafactor.py:329]: A nicer error message?

File: torch/optim/adafactor.py (hunk 6 above)
Comment (albanD) [reply torch/optim/adafactor.py:315]: Actually maybe a comment here would be nice to specify the invariant on which one is not None based on params.dim()

File: torch/optim/adafactor.py [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+                torch.norm(grad, dim=-1, keepdim=True).square_().div_(grad.size(-1))
+            )
+            row_var.lerp_(row_mean, 1 - beta2_t)
+            # same as (g * g).mean(dim=-2) w/o materializing an intermediate size g
+            col_mean = (
+                torch.norm(grad, dim=-2, keepdim=True).square_().div_(grad.size(-2))
+            )
+            col_var.lerp_(col_mean, 1 - beta2_t)
+            var_estimate = row_var @ col_var
+            var_estimate.div_(row_var.mean(dim=-1, keepdim=True).clamp_(min=eps1))
```
Comment (albanD) [inline torch/optim/adafactor.py:370]:
```suggestion
            var_estimate.div_(row_var.mean(dim=-2, keepdim=True).clamp_(min=eps1))
```

Why isn't this supposed to be this one? You already reduced dim=-1 above, you want to reduce the other dimension here?

File: torch/optim/adafactor.py [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+            update = var_estimate.sqrt_().clamp_(min=eps1).reciprocal_()
+        else:
+            grad_squared_eps = grad * grad + eps1
+            variance.lerp_(grad_squared_eps, 1 - beta2_t)
+            update = variance.rsqrt()
```
Comment (albanD) [inline torch/optim/adafactor.py:375]: Should we apply the same eps1 treatment we did for the other branch to this one?

File: torch/optim/adafactor.py [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+    if not torch._utils.is_compiling() and not all(
+        isinstance(t, torch.Tensor) for t in state_steps
+    ):
+        raise RuntimeError(
+            "`state_steps` argument must contain a list of singleton tensors"
+        )
+
+    func = _single_tensor_adafactor
+
+    func(
```
Comment (albanD) [inline torch/optim/adafactor.py:417]: I guess there is a good reason for making this a variable instead of just calling it?

File: torch/testing/_internal/common_optimizers.py [hunk 11]
```diff
@@ -119,7 +120,7 @@ def __init__(
         ),
         # A subset of the global-cliquey flags (fused, foreach, differentiable) the optimizer
         # supports. See NOTE: [optimizer kwarg categories] for what global-cliquey means.
-        supported_impls: Tuple[str] = ("foreach", "differentiable"),
+        supported_impls: Tuple[str, ...] = ("foreach", "differentiable"),
```
Comment (albanD) [inline torch/testing/_internal/common_optimizers.py:123]: enable typing for this file to make sure we don't regress these?

File: torch/optim/adafactor.py [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+            &\hspace{5mm}\textbf{else}                                                           \\
+            &\hspace{10mm}G_t           \leftarrow   \nabla_{\theta} f_t (\theta_{t-1})          \\
+            &\hspace{5mm}\widehat{\beta}_{2_t} \leftarrow 1 - t^{\tau}                           \\
+            &\hspace{5mm}\rho_t         \leftarrow min(lr, \frac{1}{\sqrt{t}})                   \\
+            &\hspace{5mm}\alpha_t       \leftarrow max(\epsilon_2,
+                \text{RMS}(\theta_{t-1}))\rho_t                                                  \\
+            &\hspace{5mm}\theta_t       \leftarrow \theta_{t-1} - \gamma \lambda \theta_{t-1}    \\
+            &\hspace{5mm}\textbf{if} \: \text{dim}(G_t) > 1:                                     \\
+            &\hspace{10mm}R_t           \leftarrow \widehat{\beta}_{2_t}R_{t-1}+
+                (1-\widehat{\beta}_{2_t})(G_t \odot G_t + 1_n \cdot 1^\top_m) \cdot 1_m          \\
```
Comment (albanD) [inline torch/optim/adafactor.py:195]: You don't actually add 1 element-wise here, you should remove the 1_n \cdot 1^\top_m
It went away when moving eps1

Same for the next line

File: torch/optim/adafactor.py (hunk 9 above)
Comment (albanD) [reply torch/optim/adafactor.py:375]: That will allow you to share "update" computation between the two branches as well

File: torch/optim/adafactor.py [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+            &\hspace{10mm}C_t           \leftarrow \widehat{\beta}_{2_t}C_{t-1}+
+                (1-\widehat{\beta}_{2_t}) 1^\top_n \cdot (G_t \odot G_t + 1_n \cdot 1^\top_m)    \\
+            &\hspace{10mm}\widehat{V}_t \leftarrow
+                \frac{R_t \cdot C_t}{max(1^\top_n \cdot R_t, \epsilon_1)}                        \\
+            &\hspace{5mm}\textbf{else}                                                           \\
+            &\hspace{10mm}\widehat{V}_t \leftarrow \widehat{\beta}_{2_t}\widehat{V}_{t-1}+
+                (1-\widehat{\beta}_{2_t}) 1^\top_n \cdot (G_t \odot G_t + \epsilon_1 1_n)        \\
```
Comment (albanD) [inline torch/optim/adafactor.py:202]: There is no "1^\top_n" here right? You don't actually do any reduction

File: torch/optim/adafactor.py [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+)
+
+__all__ = ["Adafactor", "adafactor"]
+
+
+class Adafactor(Optimizer):
+    def __init__(
+        self,
+        params: ParamsT,
+        lr: Union[float, Tensor] = 1e-2,
```
[the remark this answers is not in the document]
Comment (janeyx99) [reply torch/optim/adafactor.py:21]: yes, also because the LR is not used as it typically is used. Read the accompanying note in the docs!

File: torch/optim/adafactor.py (hunk 13 above)
Comment (janeyx99) [reply torch/optim/adafactor.py:202]: yes good catch

File: torch/optim/adafactor.py (hunk 8 above)
Comment (janeyx99) [reply torch/optim/adafactor.py:370]: It should be -2 (the -1 was leftover from before I used keepdim), and so it makes sense for it to be -2. However, it doesn't seem to matter. I'll verify on my end that the numbers are still similar.

File: torch/optim/adafactor.py (hunk 10 above)
Comment (janeyx99) [reply torch/optim/adafactor.py:417]: Just cuz in the future when we add multi_tensor this is set up for that. But maybe that's not worth doing.

File: torch/optim/adafactor.py (hunk 9 above)
Comment (janeyx99) [reply torch/optim/adafactor.py:375]: sure

File: torch/testing/_internal/common_optimizers.py (hunk 11 above)
Comment (janeyx99) [reply torch/testing/_internal/common_optimizers.py:123]: Maybe later, the entire torch/testing/_internal directory is excluded from typing and it isn't obvious how to just bring this one file back with `pytorch/mypy.ini`

File: torch/optim/_adafactor.py [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+            var_estimate = row_var @ col_var
+            var_estimate.div_(row_var.mean(dim=-2, keepdim=True).clamp_(min=eps1))
+        else:
+            assert variance is not None, "variance should be defined"
+            grad_squared = grad * grad
+            variance.lerp_(grad_squared, 1 - beta2_t)
+            # avoid writing into variance during update
+            var_estimate = variance.clone()
+
+        update = var_estimate.sqrt_().clamp_(min=eps1).reciprocal_()
```
Comment (albanD) [inline torch/optim/_adafactor.py:386]:
```suggestion
        update = var_estimate.clamp_(min=eps1).rsqrt_()
```
Would this work? It might be faster

File: torch/optim/_adafactor.py [hunk 16]
```diff
@@ -0,0 +1,444 @@
...
+            &\textbf{initialize} : \: R_0 \leftarrow 0 \text{ (second moment row factor)},       \\
+            &\hspace{23mm} \: C_0 \leftarrow 0 \text{ (second moment col factor)},               \\
+            &\hspace{23mm} \: \widehat{V}_0 \leftarrow 0 \text{ (second moment for vectors)}     \\[-1.ex]
+            &\rule{110mm}{0.4pt}                                                                 \\
+            &\textbf{for} \: t=1 \: \textbf{to} \: \ldots \: \textbf{do}                         \\
+
+            &\hspace{5mm}\textbf{if} \: \textit{maximize}:                                       \\
+            &\hspace{10mm}G_t           \leftarrow   -\nabla_{\theta} f_t (\theta_{t-1})         \\
+            &\hspace{5mm}\textbf{else}                                                           \\
+            &\hspace{10mm}G_t           \leftarrow   \nabla_{\theta} f_t (\theta_{t-1})          \\
```
Comment (albanD) [inline torch/optim/_adafactor.py:186]: This needs to be updated to match the new eps1 behavior

File: torch/optim/_adafactor.py [hunk 17]
```diff
@@ -0,0 +1,444 @@
...
+            parameter scaling. (default: (None, 1e-3))
+        d (float, optional): the clipping threshold, used to avoid larger-than-desired
+            updates.
+        weight_decay (float, optional): weight decay coefficient (default: 1e-2)
+        {_maximize_doc}"""
+    + r"""
+    .. Note::
+        The implementation of Adafactor subtly differs from Shazeer, Noam, and Mitchell Stern
+        and implementations in some other frameworks with its use of learning rate and
+        :math:`epsilon_1`.
```
Comment (albanD) [inline torch/optim/_adafactor.py:240]:
```suggestion
        :math:`\epsilon_1`.
```
here and below?

File: torch/optim/_adafactor.py (hunk 15 above)
Comment (janeyx99) [reply torch/optim/_adafactor.py:386]: It's def faster, but then we'd be square rooting eps1 --> is that..fine?

File: torch/optim/_adafactor.py (hunk 16 above)
Comment (janeyx99) [reply torch/optim/_adafactor.py:186]: oi! yes

File: torch/optim/_adafactor.py (hunk 15 above)
Comment (janeyx99) [reply torch/optim/_adafactor.py:386]: I suppose we can do whatever we want here

Comment (janeyx99) [reply torch/optim/_adafactor.py:386]: hmm in some of our other optimizers (Adam, Adagrad), the epsilon is kept unfettered. Only in Adadelta do we put the epsilon under a square root. I'm not sure what's convention here vs  whether this is something that's slidable if it's faster.

**Review by albanD — APPROVED** [review]
Looks good to me ! :)
